oqsprov/oqsx.h

```diff
@@ -19,29 +19,83 @@
 #  include <openssl/core.h>
 #  include <openssl/e_os2.h>
 
+/* Extras for OQS extension */
+
+#define ENCODE_UINT32(pbuf, i)  (pbuf)[0]   = (unsigned char)((i>>24) & 0xff); \
+                                (pbuf)[1] = (unsigned char)((i>>16) & 0xff); \
+                                (pbuf)[2] = (unsigned char)((i>> 8) & 0xff); \
+                                (pbuf)[3] = (unsigned char)((i    ) & 0xff);
+#define DECODE_UINT32(i, pbuf)  i  = ((uint32_t) (pbuf)[0]) << 24; \
+                                i |= ((uint32_t) (pbuf)[1]) << 16; \
+                                i |= ((uint32_t) (pbuf)[2]) <<  8; \
+                                i |= ((uint32_t) (pbuf)[3]);
+
+#define ON_ERR_SET_GOTO(condition, ret, code, gt) \
+    if ((condition)) {                            \
+        printf("ON_ERR_CONDITION: %d, setting code: %d\n", condition, code); fflush(stdout);   \
+        (ret) = (code);                           \
+        goto gt;                                  \
+    }
+
+#define ON_ERR_GOTO(condition, gt) \
+    if ((condition)) {                        \
+        printf("ON_ERR_CONDITION: %d\n", condition); fflush(stdout);   \
+        goto gt;                              \
+    }
+
+#define ECP_NAME(secbits, oqsname) \
+    (secbits == 128 ? "p256_" #oqsname "" : \
+     secbits == 192 ? "p384_" #oqsname "" : \
+                      "p521_" #oqsname "")
+#define ECX_NAME(secbits, oqsname) \
+    (secbits == 128 ? "x25519_" #oqsname "" : \
+                        "x448_" #oqsname "")
+
 typedef struct prov_oqs_ctx_st {
     const OSSL_CORE_HANDLE *handle;
     OSSL_LIB_CTX *libctx;         /* For all provider modules */
 //    BIO_METHOD *corebiometh; // for the time being, do without BIO_METHOD
 } PROV_OQS_CTX;
 
-
 PROV_OQS_CTX *oqsx_newprovctx(OSSL_LIB_CTX *libctx, const OSSL_CORE_HANDLE *handle);
 void oqsx_freeprovctx(PROV_OQS_CTX *ctx);
 # define PROV_OQS_LIBCTX_OF(provctx) (((PROV_OQS_CTX *)provctx)->libctx)
 
 #include "oqs/oqs.h"
 
+struct oqsx_kex_info_st {
+    int nid_kex;
+    int nid_kex_crv;
+    int raw_key_support;
+    size_t kex_length_public_key;
+    size_t kex_length_private_key;
+    size_t kex_length_secret;
+};
+
+typedef struct oqsx_kex_info_st OQS_KEX_INFO;
+
+typedef struct {
+    OQS_KEM *kem;
+    EVP_PKEY_CTX *kex;
+    EVP_PKEY *kexParam;
+    OQS_KEX_INFO kex_info;
+} OQS_HYB_KEM;
```

Comment:
Can you explain the rationale for placing the OQS_KEM pointer to different structs in different uses (either it's in OQS_PRIMITIVE or in OQS_HYB_KEM if I get it right)? That difference also necessitated changes in the OQS-only KEM provider (as you kept it separate). Generally, why did you create a fully separate KEM provider for hybrid instead of adding hybrid logic to the existing one? This will in the future necessitate changes to two files instead of one, e.g., if the OSSL APIs change. Wouldn't it have been better to define an OQS_PROV_KEY struct, featuring an OQS_KEY element (either SIG or KEM) and one (or several, depending on number of classic keys supported) classic EVP structs/pointers?

Reply:
The rationale was to define `OQS_PRIMITIVE` to contain only the contexts needed for each case: Signature -> OQS_SIG, KEM -> OQS_KEM, Hybrid KEM -> OQS_KEM + "Classical EVP structures. Having separate KEM and hybrid KEM providers (and files) was mainly for logical separation. Good point about merging hybrid logic with the normal KEM logic, some code duplication could also be avoided that way. I will do some refactoring that way and think about the structures.

Reply:
Thanks very much! Don't be shy changing `oqs_kem.c` (or any other file) upside-down. I'm still not on top of the encode APIs, so will add whatever I may eventually come up with on top of this PR when merged.

Reply:
Thanks. All KEM/hybrid KEM logic is now in oqs_kem.c. There are internal functions for OQS KEMs (`oqs_qs_kem..`) and for EVP-based KEMs (`oqs_evp_kem..`). Then there are "hybrid" functions that combine the two mentioned types.
OQSX_KEY now holds pointers to pointers to different subkeys (`void **privkey_ptr`, `void **pubkey_ptr`). In the usual KEM case, there is only one. In the hybrid case there are two. If there is a need for further keys (e.g. for hybrids with more than 2 schemes), this will be easy to add.
As you suggested, I also refactored the structs holding the OQS and EVP contexts. Previously they were in OQS_PRIMITIVE. Now there is `OQSX_PROVIDER_CTX` that holds the OQS_KEM/OQS_SIG and the EVP contexts.
Hypotetically, OQSX_PROVIDER_CTX could be extended with further context types (e.g. for using other OSSL API).